domains/client/domain-operator/src/fetch_domain_bootstrap_info.rs

```diff
@@ -39,13 +46,48 @@ where
         .runtime_api()
         .domain_instance_data(best_hash, self_domain_id)?
     {
+        let domain_best_number = consensus_client
+            .runtime_api()
+            .domain_best_number(best_hash, self_domain_id)?
+            .unwrap_or_default();
+
+        // The `domain_best_number` is the expected best domain block after the operator has
+        // processed the consensus block at `best_hash`. If `domain_best_number` is not zero
+        // and the domain chain is not started, the domain block `0..domain_best_number` are
+        // missed, we can not preceed running as a domain node because:
+        //
+        // - The consensus block and state that derive the domain block `0..domain_best_number`
+        //    may not available anymore
+        //
+        // - There may be domain runtime upgrade in `0..domain_best_number` which will result
+        //    in inconsistent `raw_genesis`
+        if !is_domain_started && !domain_best_number.is_zero() {
+            return Err(
+                "Unable to restart the existing consensus node as a domain node"
+                    .to_string()
+                    .into(),
+            );
+        }
+
         return Ok(BootstrapResult {
             domain_instance_data,
             domain_created_at,
             imported_block_notification_stream,
         });
     }
 
+    // The domain instance data is not found in the consensus chain while the domain chain
+    // is already started meaning the domain chain is running ahead of the consensus chain,
+    // which is unexpected as the domain chain is derived from the consensus chain.
+    if is_domain_started {
+        return Err(
+            "The domain chain is ahead of the consensus chain, possibly uncleaned `domains`
+            folder from the last run"
```

Comment:
It might help to name both inconsistent folders here

```suggestion
            "The domain chain is ahead of the consensus chain, inconsistent `db` and `domains`
            folders from the last run"
```